File: labs/lab3/C++/Vector.cpp
#include "Vector.h"
#include <cmath>

// Конструктор за замовчуванням
Vector::Vector()
    : x(0.0), y(0.0)
{
}

// Конструктор з параметрами
Vector::Vector(double xVal, double yVal)
    : x(xVal), y(yVal)
{
}

// Конструктор копіювання
Vector::Vector(const Vector& other)
{
    x = other.x;
    y = other.y;
}

// 1) length() без параметрів
double Vector::length() const
{
    return std::sqrt(x * x + y * y);
}

// 2) length(double scale)
double Vector::length(double scale) const
{
    // Якщо уявити, що вектор масштабуємо на scale,
    // то довжина змінюється у scale разів.
    return this->length() * scale;
}

// Гетери
double Vector::getX() const { return x; }
double Vector::getY() const { return y; }

// Перевантажений оператор +
Vector Vector::operator+(const Vector& rhs) const
{
    return Vector(x + rhs.x, y + rhs.y);
}

// Перевантажений оператор /
Vector Vector::operator/(double val) const
{
    // Не перевіряємо val на 0 для спрощення
    return Vector(x / val, y / val);
}

File: labs/lab3/C#/Vector.cs
namespace Lab1_3_Vectors
{
    public class Vector
    {
        private double x;
        private double y;

        // Конструктор за замовчуванням
        public Vector()
        {
            x = 0.0;
            y = 0.0;
        }

        // Конструктор з параметрами
        public Vector(double xVal, double yVal)
        {
            x = xVal;
            y = yVal;
        }

        // Конструктор копіювання
        public Vector(Vector other)
        {
            x = other.x;
            y = other.y;
        }

        // Метод обчислення довжини
        public double Length()
        {
            return System.Math.Sqrt(x * x + y * y);
        }

        // Гетери (щоб зчитати координати)
        public double GetX() { return x; }
        public double GetY() { return y; }

        // Перевантаження оператора +
        public static Vector operator +(Vector a, Vector b)
        {
            return new Vector(a.x + b.x, a.y + b.y);
        }

        // Перевантаження оператора / (ділення на число)
        public static Vector operator /(Vector a, double val)
        {
            // Для спрощення не робимо перевірку на val == 0
            return new Vector(a.x / val, a.y / val);
        }
    }
}
